aesd-char-driver/main.c

```diff
@@ -186,14 +191,94 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
     mutex_unlock(&(buffer->lock));
     return retval;
 }
+// ------------------------------aesd_adjust_file_offset-------------------------------------------
+
+static long aesd_adjust_file_offset(struct file *filp, unsigned int write_cmd, unsigned int write_cmd_offset)
+{
+    int i, tempfilepos, retval = 0;
+
+    struct aesd_dev *buffer = filp->private_data;
+
+    mutex_lock(&aesd_device.lock);
+
+    if (write_cmd > (AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED - 1))
+    {
+        retval = -EINVAL;
+        goto exit;
+    }
+
+    // Check if offset is valid
+    if (write_cmd_offset >= buffer->data_buffer.entry[write_cmd].size)
+    {
+        retval = -EINVAL;
+        goto exit;
+    }
+    tempfilepos = 0;
+    for (i = 0; i < write_cmd; i++)
+    {
+        if (buffer->data_buffer.entry[i].size == 0)
+        {
+            retval = -EINVAL;
+            goto exit;
+        }
+        tempfilepos += buffer->data_buffer.entry[i].size;
+    }
+    tempfilepos += write_cmd_offset;
+    filp->f_pos = tempfilepos;
+
+exit:
+    mutex_unlock(&aesd_device.lock);
+    return retval;
+}
+// ------------------------------aesd_ioctl-------------------------------------------
+
+long aesd_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
+{
+    long retval = 0;
+    struct aesd_seekto seek_to_pos;
+
+    if ((_IOC_TYPE(cmd) != AESD_IOC_MAGIC) || (_IOC_NR(cmd) > AESDCHAR_IOC_MAXNR))
+        return -ENOTTY;
+
+    switch (cmd)
+    {
+    case AESDCHAR_IOCSEEKTO:
+        if (copy_from_user(&seek_to_pos, (const void __user *)arg, sizeof(seek_to_pos)) != 0)
+            retval = -EFAULT;
+        else
+            retval = aesd_adjust_file_offset(filp, seek_to_pos.write_cmd, seek_to_pos.write_cmd_offset);
+        break;
+
+    default:
+        retval = -ENOTTY;
+        break;
+    }
+
+    return retval;
+}
+// ------------------------------aesd_llseek-------------------------------------------
+loff_t aesd_llseek(struct file *filp, loff_t off, int whence)
+{
+    loff_t retval;
+    struct aesd_dev *dev = filp->private_data;
+    mutex_lock(&aesd_device.lock);
+    retval = fixed_size_llseek(filp, off, whence, dev->data_buffer.total_buff_size);
```

Comment:
@neohere97 think about how you could change in a way which wouldn't require you to hold the lock during `fixed_size_llseek`